### ISSUE TITLE
Support PreciPoint VMIC format

### ISSUE THREAD

**Issue by bgilbert** [openslide/openslide#168]
Apparently it's [roughly Deep Zoom inside ZIP](https://lists.andrew.cmu.edu/pipermail/openslide-users/2015-December/001164.html).

**Comment by Markus-PP** [comment]
I am ready to work on this. Has anyone else given the matter some thoughts yet ? Obviously it requires a ZIP library. After some tests, I concluded that libzip does what we need. Are there any objections against including libzip ? Will there be other slides with images in containers supported by libzip ?

**Comment by bgilbert** [comment]
Awesome, thanks for looking into this!  When writing your code, please ensure that you do not use any information obtained under a non-disclosure agreement.  If you use NDA-protected material, we will not be able to merge your changes.

libzip would work, but wouldn't be my first choice.  libgsf will read both ZIP and OLE files, which might be useful someday for Zeiss CZI.  If you do use libzip, note that version 1.0 is not yet available in many supported Linux distributions.  I'm not currently aware of any other primary formats that use ZIP containers, but I'd suggest making an `openslide-decode-gsf.c` (or `-decode-zip.c`) anyway.

Do you have sample VMIC files?

**Comment by bgilbert** [comment]
Does VMIC use ZIP64?

- libgsf < 1.14.31 (Debian through jessie, Ubuntu through vivid, RHEL 6 and 7) [will choke on ZIP64](https://bugzilla.gnome.org/show_bug.cgi?id=732209).
- libzip < 0.11 (Debian through wheezy, Ubuntu through trusty, RHEL 6 and 7) [doesn't support ZIP64](http://www.nih.at/libzip/NEWS.html), and from a quick glance at the 0.10.1 source, may also choke.

**Comment by Markus-PP** [comment]
Thanks for fast reply ! I have approval and support of the PreciPoint team so i don't see any legal issue here. I do have sample files, and I know ZIP64 is actually required. Thanks for hinting that older versions don't support it - I did test runs with libzip 1.1.2 (the newest) which works like charm on VMIC, will have to check if libzip 0.11.2 and libgsf do the job, too. Also planning to set up an openslide-decode-deepzoom.c for generalizing purposes. Will probably have more questions but not at this point.

[My distribution, OpenSuSE 13.2, has libzip 0.11.2 out-of-the-box, updateable from repo to 1.1.2; and libgsf 1.14.30, updateable to 1.14.33]

**Comment by bgilbert** [comment]
If older versions of a library simply reject ZIP64, that's fine.  If they crash or behave incorrectly (as libgsf does) we will need to conditionally enable ZIP support.

Are you able to share the sample files under [CC0](https://creativecommons.org/publicdomain/zero/1.0/)?  We strongly prefer that new drivers come with freely-redistributable samples.  (Indeed, we can't create automated test cases without them.)

Some background on the NDA stuff: we have encountered vendors who have offered to provide documentation for an OpenSlide driver, but only under restrictions which are unworkable for an open-source project.  Therefore, to avoid problems, we've been planning to require contributors to certify that they have the right to contribute the code they are submitting under OpenSlide's license.  It's the same process used in Linux kernel development: each Git commit message will need to include a `Signed-off-by:` line (as produced by `git commit -s`), which certifies that the change is submitted in accordance with the [Developer Certificate of Origin](http://developercertificate.org/).  I haven't had a chance yet to document this policy on the OpenSlide website, but it will apply to your contributions.  Please let me know if you have any questions, and thanks for your understanding.

**Comment by Markus-PP** [comment]
PreciPoint team have agreed on providing some VMIC- example files for public domain.

https://drive.google.com/file/d/0BwWPfDJXmk2gbm5ROU5TZUNxVms/view?usp=sharing
https://drive.google.com/file/d/0BwWPfDJXmk2gbUdsU1V2TGE4eHM/view?usp=sharing
https://drive.google.com/file/d/0BwWPfDJXmk2gcG5NME5zNGgwTnM/view?usp=sharing

I have written some code which is in experimental stage and somewhat working, although not ready for inclusion in the library. Shall I discuss it here or rather in the mailing list instead ?

**Comment by bgilbert** [comment]
Thanks for the slide files!  Note that public domain isn't legally valid in all countries.  Would it be possible for the samples to be licensed under CC0?  (CC0 is essentially public domain plus a permissive license for places where public domain doesn't exist.)

If your code isn't ready for a pull request, I'd suggest discussing it on the mailing list.

**Comment by Markus-PP** [comment]
Yes, of course they can be licensed under CC0, as well, since the PreciPoint team ensured they are devoid of third-party rights. Is perhaps any official statement needed ?

A commit is under way, but before doing so i'm going to send a reply on the mailing list.

**Comment by Markus-PP** [comment]
Repository https://github.com/Markus-PP/openslide-vmic has a functional driver.

**Comment by bgilbert** [comment]
@Markus-PP For now, if you could post an official statement in this thread (or send it to me privately), I could add the samples to openslide-testdata.  That would allow VMIC tests to be added to the test suite.

Eventually we'll have a web form for collecting CC0 signoffs, and at that point I'll ask you to complete it.

**Comment by Markus-PP** [comment]
**I hereby confirm that the PreciPoint team agrees to release the above mentionend 3 slides under CC0 and that, to the best of my knowledge, they are devoid of any third party rights.**

**Comment by bgilbert** [comment]
Thank you very much!

I just noticed that each sample includes a copy of OpenSeadragon, which can't be licensed under CC0 because it is subject to third-party copyright.  OpenSeadragon is BSD-licensed and the license information is properly included in the slide files, so we can redistribute the samples, but I'll need to mark them as a mix of CC0 and BSD.

In each of the inner ZIP files, there is a small number of members with a `.jpg.bmp` extension.  What is the purpose of those members?

**Comment by bgilbert** [comment]
Also:

- Can you easily provide any additional descriptive metadata for each of the files?  (Stain, type of tissue, etc.)  It would be useful for the textual description of each slide ([example](http://openslide.cs.cmu.edu/download/openslide-testdata/Leica/)).
- Do you or PreciPoint want to be credited for the samples, and if so, how should we credit you?  (See above example.)
- Does your software care about the names of slide files?  In other words, if I rename the files to `<something>.vmic`, will that cause any problems?

**Comment by Markus-PP** [comment]
To the first question: I guess, at the moment it's probably best to remove the seadragon references, because they are not needed for any of the viewer applications. The original goal was to provide the customer a way to easily unpack and view the slide without extra software. The bogus jpg.bmp files, if I remember correctly, were created during testing and they have no purpose. I'll talk with the team next week to know for sure, Then, I can also fetch the textual description, together with the seadragon-free slides.
Any credits should go to PreciPoint, if possible.
The VMIC driver as implemented here does not care about any part of the file name, nor does the viewer application, so yes they can be renamed to anything.

**Comment by bgilbert** [comment]
Great, I'll credit PreciPoint and rename the files.

We should avoid modifying the samples after they are written by your software; otherwise they won't be accurate samples.  If current versions of your software no longer embed OpenSeadragon, then I'd be in favor of generating some new scans.  Otherwise, let's use unmodified files and just list both licenses.  (Mixed licensing is not uncommon in open-source projects.)

**Comment by kwiechen** [comment]
The download links do not work anymore- is it possible to get the CC0 and BSD .vmic example files?

**Comment by kasper-cancilico** [comment]
Reviving this after a long gap. I'd like to see .vmic readable in OpenSlide-based tools, so I went back through the history here to check whether there's still appetite to get it in.

Where things stand, as far as I can tell:

- @Markus-PP wrote a working driver back in 2016–2019 in his fork ([Markus-PP/openslide-vmic](https://github.com/Markus-PP/openslide-vmic), `vmic` branch): an `openslide-vendor-precipoint.c` plus an `openslide-decode-zip.c` helper, libzip-based. It never landed in core and targets the old 3.x architecture.
- The libzip question looked mostly settled at the time: libzip handles the ZIP64 that VMIC needs, and the openslide-bin libzip PR (#12) was merged. I don't see that support in openslide-bin now, so it looks like it was dropped along the way, and OpenSlide has since moved to 4.x/meson.
- The CC0 sample links from back then are dead (as noted here in 2025).

A few questions for the current maintainers (@bgilbert and whoever else is active):

1. Is there still interest in a PreciPoint / DeepZoom-in-ZIP driver for OpenSlide 4.x?
2. Would libzip still be your preference, or has the thinking changed (minizip-ng, libgsf)? The zip handling sits behind a small helper, so the library is easy to swap.
3. Anything you'd want up front, like DCO sign-off or fresh CC0 samples?

If it's welcome, I'm happy to take the port to 4.x, building on @Markus-PP's existing driver. Just wanted to check there's appetite before anyone sinks time into it.
